pam/go-loader/module.c

```diff
@@ -0,0 +1,142 @@
+/* A simple PAM wrapper for GO based pam modules
+ *
+ * Copyright (C) 2023 Marco Trevisan
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General
+ * Public License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Marco Trevisan <marco.trevisan@canonical.com>
+ */
+
+#include <dlfcn.h>
+#include <limits.h>
+#include <security/pam_modules.h>
+#include <security/pam_ext.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+/* When a Go shared library is loaded from C, go starts various goroutine
+ * (as init() at first) and if the loading code is then performing a fork
+ * we end up having an undefined behavior and very likely, deadlocks.
+ *
+ * To avoid this we need to ensure that the module is called after the loading
+ * application has forked, and this for sure has happened when the application
+ * calls the PAM functions that a module exposes.
+ *
+ * As per this, here we re-implement the PAM modules functions and for each
+ * action we load the module if it has not been loaded already, otherwise we
+ * dload it and we redirect each call to the loaded library.
+ */
+
+typedef int (*PamHandler)(pam_handle_t *,
+                          int          flags,
+                          int          argc,
+                          const char **argv);
+
+static void
+on_go_module_removed (pam_handle_t *pamh,
+                      void         *go_module,
+                      int           error_status)
+{
+  void (*go_pam_cleanup) (void);
+  *(void **) (&go_pam_cleanup) = dlsym (go_module, "go_pam_cleanup_module");
+  if (go_pam_cleanup)
+    go_pam_cleanup ();
+
+  dlclose (go_module);
+}
+
+static void *
+load_module (pam_handle_t *pamh,
+             const char   *module_path)
+{
+  void *go_module;
+
+  if (pam_get_data (pamh, "go-module", (const void **) &go_module) == PAM_SUCCESS)
+    return go_module;
+
+  go_module = dlopen (module_path, RTLD_LAZY);
+  if (!go_module)
+    return NULL;
+
+  pam_set_data (pamh, "go-module", go_module, on_go_module_removed);
```

Comment:
So, I think you are using pam_set_data to get the associated cleanup function. What is the advantage compared to having a setter function when we associate it to NULL?

My bet is that your goal is when the PAM transaction ends, all values are reset to NULL and so the pam go module is unloaded, is that the case? (At worst, we can have this done once we are get to the last stage)

Reply:
> So, I think you are using pam_set_data to get the associated cleanup function. What is the advantage compared to having a setter function when we associate it to NULL?

Not sure I understand what you mean :-)

> My bet is that your goal is when the PAM transaction ends, all values are reset to NULL and so the pam go module is unloaded, is that the case? 

Yes, that's the point here.

Reply:
Excellent! (I was comparing this with an internal setter function), but yeah, if pam automatically reset that to NULL which ends up unloading the module, this is great!